### PR TITLE
fix: decode uri right before routing #3356

### PR DESCRIPTION
**Pull request by mav-rik**
#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mav-rik** [comment]
fix for #3356

File: fastify.js [hunk 1]
```diff
@@ -672,13 +672,19 @@ function validateSchemaErrorFormatter (schemaErrorFormatter) {
   }
 }
 
-function wrapRouting (httpHandler, { rewriteUrl, logger }) {
-  if (!rewriteUrl) {
-    return httpHandler
+function safeDecodeURI (url, logger) {
+  try {
+    return decodeURI(url)
```
Comment (mcollina) [inline fastify.js:677]: please use https://github.com/delvedor/fast-decode-uri-component.

**Review by mcollina — COMMENTED** [review]
I'm unsure about this change.

find-my-way is applying this decoding _already_: https://github.com/delvedor/find-my-way/blob/c561d220e6c0c93b76b72a7461dcdc8620158ba2/index.js#L466 (and other places).

I might have a feeling we should do it there instead, this looks a lot like a bug in find-my-way.

File: fastify.js (hunk 1 above)
Comment (mav-rik) [reply fastify.js:677]: `fast-decode-uri-component` behaves differently from `decodeURI`. Please see my [comment](https://github.com/fastify/fastify/issues/3356#issuecomment-933778630).

File: test/hooks.test.js [hunk 2]
```diff
@@ -714,7 +714,7 @@ test('onRoute hook should able to change the route url', t => {
   fastify.register((instance, opts, done) => {
     instance.addHook('onRoute', (route) => {
       t.equal(route.url, '/föö')
-      route.url = encodeURI(route.url)
+      route.url = '/föö2'
```
Comment (mcollina) [inline test/hooks.test.js:717]: Why this change?

Comment (mav-rik) [reply test/hooks.test.js:717]: Since we apply `decodeURI` to each url right after we get a request it becomes unnecessary and useless to `encodeURI` for routes. The test name was `onRoute hook should able to change the route url` so I kept the test to be doing what's expected. The `onRoute` hook does change the route url and the test makes sure that this works.

But I think I am going to close this PR because it looks like it has to be fixed in `find-my-way` lib. I also noticed that `fast-decode-uri-component` works faster only when there is at least 25% of malformed URIs. When there is no malformed URIs the standard `decodeURIComponent` works up to 2 times faster (especially if you add a precondition ```if (uri.indexOf('%') < 0) return uri``` before calling the `decodeURIComponent`)

**Comment by mav-rik** [comment]
After further investigation I made a conclusion that the fix must be applied to `find-my-way` lib.
